pywbem_mock/_instancewriteprovider.py

```diff
@@ -270,367 +209,342 @@ def __init__(self, cimrepository=None):
+            This object is a deep copy of the original client parameter, and may
+            be modified by the provider as needed, before storing it in the
+            CIM repository.
 
-            The namespace of the ``path`` attribute must be `None` or the
-            same as the namespace parameter.
+            The `classname` attribute of this object will specify the
+            creation class for the new instance, in any lexical case.
 
-            The `properties` attribute of this object specifies initial
-            property values for the new CIM instance.
+            The `properties` attribute of this object will specify the
+            initial property values for the new CIM instance, with property
+            names in any lexical case. Key properties may or may not be
+            included.
 
-            Instance-level qualifiers have been deprecated in CIM, so any
-            qualifier values specified using the `qualifiers` attribute
-            of this object are ignored.
+            The `path` attribute of this object will be `None`.
+
+            The `qualifiers` attribute of this object, if non-empty, should
+            be ignored by the provider, because instance-level qualifiers have
+            been deprecated in CIM.
 
         Returns:
 
             :class:`~pywbem.CIMInstanceName`: Instance path of the new CIM
             instance.
 
         Raises:
-
-            :exc:`~pywbem.CIMError`: (CIM_ERR_INVALID_NAMESPACE)
-            :exc:`~pywbem.CIMError`: (CIM_ERR_INVALID_PARAMETER)
-            :exc:`~pywbem.CIMError`: (CIM_ERR_INVALID_CLASS)
-            :exc:`~pywbem.CIMError`: (CIM_ERR_ALREADY_EXISTS)
+            :exc:`~pywbem.CIMError`: The provider may raise CIMError with any
+              status code, and typically raises:
+              - CIM_ERR_INVALID_PARAMETER
+              - CIM_ERR_ALREADY_EXISTS
         """
 
-        # Parameter types are already checked by WBEMConnection operation
-        assert isinstance(namespace, six.string_types)
-        assert isinstance(NewInstance, CIMInstance)
-
-        new_instance = NewInstance.copy()
+        # Get the creation class with all exposed properties and qualifiers.
+        # Since the existence of the class has already been verified, this
+        # will always succeed.
+        class_store = self.cimrepository.get_class_store(namespace)
+        creation_class = class_store.get(NewInstance.classname, copy=False)
+
+        # This default provider adjusts the lexical case of the property names
+        # in the new instance to match the property definitions in the creation
+        # class. A user-defined provider may or may not do that.
+        for inst_pn in NewInstance.properties:
+            inst_prop = NewInstance.properties[inst_pn]
+            cls_pn = creation_class.properties[inst_pn].name
+            if inst_pn != cls_pn:
+                inst_prop.name = cls_pn  # modifies NewInstance
```

Comment:
DISCUSSION:

The adjustment of lexical case could be moved up to the dispatcher, if we believe this is good practice. If we think this is a don't care, it can be removed from here as well.

Reply:
We agreed to go for that.

Reply:
DONE in latest commit.